### PR TITLE
Add support for HiDPI displays

### PR DESCRIPTION
**Pull request by xjuan**
Webkit2Gtk uses a 1 to 1 backing store relation for canvas element which
means there is not native support to display HiDPI images in it.
The only way to do so is to create a canvas bigger and then scale it back
with CSS zoom property.

https://phabricator.endlessm.com/T22183

### CODE REVIEW

File: src/maxwell-web-view.js [hunk 1]
```diff
@@ -27,13 +27,20 @@ let children = [];             /* List of children */
...
     let positions = null;
 
     for (let i = 0, len = children.length; i < len; i++) {
         let child = children[i];
         let child_rect = child.maxwell.rect;
         let rect = child.getBoundingClientRect();
 
+        /* FIXME: getBoundingClientRect() returns a wrong value when zoom is set */
+        if (scale !== 1) {
+            rect.x = (rect.x - window.scrollX * (scale - 1)) / scale;
```
Comment (ptomato) [inline src/maxwell-web-view.js]: For scale = 2, `rect.x = (rect.x - window.scrollX * 1) / 2` which seems correct
For scale = 4, `rect.x = (rect.x - window.scrollX * 3) / 4` which seems incorrect.
Are you sure about the `(scale - 1)`?

**Review by ptomato — CHANGES_REQUESTED** [review]
It mostly works for me, so if nothing else we can just merge it.

Have you seen [`CanvasRenderingContext2D.scale()`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/scale) and might that work better?

Also there is one weirdness when I test it. The canvas doesn't get painted until it is fully scrolled into the viewport. It seems like it ought to be painted when it's partially visible.

**Comment by xjuan** [comment]
putImageData() is not affected by context transformations, we could can render the image data to a offscreen canvas and use drawImage() instead, but that leaves you with a low rez canvas, because the backing store is always 1 to 1 with CSS size.
So the only way to show HiDPI is to make the canvas N times bigger and then scale it back, which works only as a side effect of an implementation detail IMO

File: src/maxwell-web-view.js (hunk 1 above)
Comment (xjuan) [reply src/maxwell-web-view.js]: Yes, it does not work without (scale-1)

rect.x is in pixels and window.scrollX is in CSS pixels so I think the problem is they forgot to multiply the scroll value by devicePixelRatio before subtracting it and that is why I need to subtract what (window.scrollX * (scale - 1))

**Pull request merged**